### PR TITLE
Support java worker resources management.

### PR DESCRIPTION
**Pull request by kingchin1218**
### This change support resources management in java worker.

i. Specify the static resources of a node.
ii. Specify the required resources of a task.

**[How to use?]**
#### 1. Specify node static resources:
(1)
Overwrite the `static_resources` filed in `ray.config.ini` file,
      then use this `./run.sh start --head ` command to start node manager.
**or**
use `./run.sh start --head --overwrite=Resource1:10,Resouces2:100` to start node manager directly. 

(2)
#### Specify the required resources of your remote method or remote actor.
```java
  @RayRemote(resources = {@ResourceItem(name = "Resource1", value = 10),
                          @ResourceItem(name = "Resource2", value = 2)})
  public static String sayHello() {
    String ret = "he";
    ret += "llo";
    return ret;
  }
```

## Related issue number
N/A

### CODE REVIEW

File: java/runtime-common/src/main/java/org/ray/spi/model/TaskSpec.java [hunk 1]
```diff
@@ -65,4 +70,20 @@ public String toString() {
     }
     return builder.toString();
   }
-}
\ No newline at end of file
+
+  private String getResourcesToString() {
```
Comment (zhijunfu) [inline java/runtime-common/src/main/java/org/ray/spi/model/TaskSpec.java]: rename to ```resourcesToString``` or ```getResourcesString```?

File: java/runtime-native/src/main/java/org/ray/spi/impl/DefaultLocalSchedulerClient.java [hunk 2]
```diff
@@ -221,12 +222,24 @@ public static ByteBuffer taskSpec2Info(TaskSpec task) {
     // resources required by this task. The index in this vector corresponds to
     // the resource type defined in the ResourceIndex enum. For example,
 
-    int[]requiredResourcesOffsets = new int[1];
-    for (int i = 0; i < requiredResourcesOffsets.length; i++) {
-      int keyOffset = 0;
-      keyOffset = fbb.createString(ByteBuffer.wrap("CPU".getBytes()));
-      requiredResourcesOffsets[i] = ResourcePair.createResourcePair(fbb, keyOffset, 0.0);
+    final String CPULiteral = "CPU";
```
Comment (zhijunfu) [inline java/runtime-native/src/main/java/org/ray/spi/impl/DefaultLocalSchedulerClient.java]: Is it the intent to update taskSpec as well as requiredResourcesOffsets?

Comment (kingchin1218) [reply java/runtime-native/src/main/java/org/ray/spi/impl/DefaultLocalSchedulerClient.java]: Yep

File: java/runtime-native/src/main/java/org/ray/runner/RunManager.java [hunk 3]
```diff
@@ -350,10 +346,18 @@ public void startRayProcesses() {
       startObjectStore(0, info, params.working_directory + "/store",
           params.redis_address, params.node_ip_address, params.redirect, params.cleanup);
 
+      Map<String, Double> staticResources = getResourcesFromString(params.static_resources);
+      if (!staticResources.containsKey("CPU")) {
+        staticResources.put("CPU", (double) params.num_cpus[0]);
+      }
+      if (!staticResources.containsKey("GPU")) {
+        staticResources.put("GPU", (double) params.num_gpus[0]);
+      }
```
Comment (raulchen) [inline java/runtime-native/src/main/java/org/ray/runner/RunManager.java]: should we deprecate `num_cpus/num_gpus` and only use `static_resources` (at least for raylet)?

Comment (kingchin1218) [reply java/runtime-native/src/main/java/org/ray/runner/RunManager.java]: @raulchen Can't agree more.

File: java/runtime-common/src/main/java/org/ray/spi/LocalSchedulerProxy.java [hunk 4]
```diff
@@ -118,6 +121,16 @@ private void doSubmit(RayInvocation invocation, UniqueID taskId,
     scheduler.submitTask(task);
   }
 
+  private Map<String, Double> getResourcesMap(ResourceItem[] resourceArray) {
+    Map<String, Double> resourceMap = new HashMap<>(16);
```
[the remark this answers is not in the document]
Comment (kingchin1218) [reply java/runtime-common/src/main/java/org/ray/spi/LocalSchedulerProxy.java]: @raulchen 
it's init_size.
If I call ctor without argument, ide advices me to add a init_size for it.

File: java/ray.config.ini [hunk 5]
```diff
@@ -62,9 +62,9 @@ deploy = false
 
 onebox_delay_seconds_before_run_app_logic = 0
 
-use_raylet = false
+use_raylet = true
```
Comment (raulchen) [inline java/ray.config.ini]: +1 for enabling use_raylet by default

File: java/ray.config.ini [hunk 6]
```diff
@@ -62,9 +62,9 @@ deploy = false
...
 
-static_resources = CPU:4,GPU:0
+static_resources = CPU:4,GPU:0,ResourceA:10,ResourceB:8
```
Comment (raulchen) [inline java/ray.config.ini]: If there're spaces between items, e.g., `CPU:4, GPU:0, ResourceA:10,   ResourceB:8`, can we still parse it correctly?

Comment (kingchin1218) [reply java/ray.config.ini]: It will parse the string correctly now.

**Comment by kingchin1218** [comment]
@zhijunfu  @raulchen 
I have addressed comments. Could u plz tak a look again?

File: java/runtime-native/src/main/java/org/ray/runner/RunManager.java [hunk 7]
```diff
@@ -852,6 +860,32 @@ private AddressInfo startObjectManager(int index, AddressInfo info, String workD
...
+
+  String resourcesToString(Map<String, Double> resources) {
+    StringBuilder builder = new StringBuilder();
+    int count = 1;
+    for (Map.Entry<String, Double> entry : resources.entrySet()) {
+      if (count != resources.size()) {
+        builder.append(entry.getKey()).append(",").append(entry.getValue()).append(",");
+      } else {
+        builder.append(entry.getKey()).append(",").append(entry.getValue());
+      }
```
Comment (raulchen) [inline java/runtime-native/src/main/java/org/ray/runner/RunManager.java]: better this way:
```java
builder.append(entry.getKey()).append(",").append(entry.getValue())
if (count != resources.size()) {
  builder.append(",");
}
```

File: java/test/src/main/java/org/ray/api/test/ResourcesManagementTest.java [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+  @Test
+  public void testMethods() {
+    if (RayRuntime.getParams().use_raylet) {
+      // This is a case that can satisfy required resources.
+      RayObject<Integer> result1 = Ray.call(ResourcesManagementTest::echo1, 100);
+      Assert.assertEquals(100, (int) result1.get());
+
+      // This is a case that can't satisfy required resources.
+      final RayObject<Integer> result2 = Ray.call(ResourcesManagementTest::echo2, 200);
+      WaitResult<Integer> waitResult = Ray.wait(result2, 3 * 1000);
```
Comment (raulchen) [inline java/test/src/main/java/org/ray/api/test/ResourcesManagementTest.java]: is it necessary to wait for so long? it will slow down the test.

**Pull request closed without merging**